### PR TITLE
Add handcrafting transaction docs rough cut

### PR DESCRIPTION
**Pull request by sbellem**
**Note to reviewers**:

* The plan of this PR is to provide a rough cut of working examples that have almost no dependence on the driver itself.
* The examples are meant to be compatible with the latest PyPI release, 0.1.2.
* Once merged, the plan is to make a maintenance release (0.1.3) for which these new docs will be available.

Subsequent work will be part of 0.2.0 such that the driver will be brought up to speed with the latest server code.

As for the examples, work will be done to remove all dependencies on the driver, such that they can be shown under the server docs,

### CODE REVIEW

File: dev.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+version: '2'
```
Comment (sohkai) [inline dev.yml]: Can this be a different PR?

File: docs/advanced-usage.rst [hunk 2]
```diff
@@ -1,3 +1,5 @@
+.. _advanced-usage:
```
Comment (sohkai) [inline docs/advanced-usage.rst]: I assume that this is basically just #187?

File: docs/handcraft.rst [hunk 3]
```diff
@@ -0,0 +1,1940 @@
...
+
+.. contents::
+    :local:
+    :depth: 1
+
+********
+Overview
+********
+
+Submitting a transaction to a BigchainDB node consist of three main steps:
```
Comment (sohkai) [inline docs/handcraft.rst]: consist*s*

File: docs/handcraft.rst [hunk 4]
```diff
@@ -0,0 +1,1940 @@
...
+
+1. Preparing the transaction payload.
+2. Fulfilling the prepared transaction payload.
+3. Sending the transaction payload via HTTPS.
+
+Step 1 and 2 can be performed offline on the client side. That is, they do not
+require any connection to any BigchainDB node.
+
+For convenience's sake some utilites are provided to prepare and fulfill a
```
Comment (sohkai) [inline docs/handcraft.rst]: sake**,**

File: docs/handcraft.rst [hunk 5]
```diff
@@ -0,0 +1,1940 @@
...
+transaction via the :class:`~.bigchaindb_driver.BigchainDB` class, and via the
+:mod:`~bigchaindb_driver.offchain` module.
```
Comment (sohkai) [inline docs/handcraft.rst]: I'd reformulate the next two paragraphs and do some merging. It'd be:

> via the :mod:`offchain` module. **For an introduction on using these utilities, see the :ref:`basic-usage` or :ref:`advanced-usage` sections.**
> 
> **The rest of this document will guide you through completing steps 1 and 2 yourself by revisiting some of the examples provided in the usage sections. We will:**..

File: dev.yml (hunk 1 above)
Comment (sbellem) [inline dev.yml]: yes, sorry: https://github.com/bigchaindb/bigchaindb-driver/pull/190

File: docs/advanced-usage.rst (hunk 2 above)
Comment (sbellem) [inline docs/advanced-usage.rst]: It was added because it is referenced to in `handcraft.rst`.

File: docs/handcraft.rst (hunk 5 above)
Comment (sbellem) [inline docs/handcraft.rst]: adressed in https://github.com/bigchaindb/bigchaindb-driver/pull/176/commits/4cf6b191bddf1479bfcae98301a194fabd3929f2

File: docs/handcraft.rst [hunk 6]
```diff
@@ -0,0 +1,2049 @@
...
+
+fulfillments
+^^^^^^^^^^^^
+At the moment, we have to distinguish between fulfillment payloads **before**
+fulfillment and **after** fulfillment. For signature-based schemes, this means,
+before signature(s), and after signature(s).
+
+Moreover payloads for ``CREATE`` operations are a bit different.
+
+.. note:: We hope to be able to simplify the payload strucutre and validation,
```
Comment (r-marques) [inline docs/handcraft.rst]: typo `strucuture`

File: docs/handcraft.rst [hunk 7]
```diff
@@ -0,0 +1,2049 @@
...
+            },
+            'uri': 'cc:4:20:bOZjTedaOgPsbYjh3QeOEQCj1o1lIvVefR71sS8egnM:96'
+        },
+        'owners_after': ['8L6ngTZ5ixuFEr1GiunrFNWtGkft4swWWArXjWJu2Uwc'],
+    }
+    
+
+fulfillments
+^^^^^^^^^^^^
+At the moment, we have to distinguish between fulfillment payloads **before**
```
Comment (r-marques) [inline docs/handcraft.rst]: I find this paragraph a bit confusing. I understand what you are trying to say.
What actually want is to construct the fulfillment object but without the signature.

One reason to do this is to avoid exposure of the private key. I may want to use some utility to generate the payloads (e.g. `bigchaindb_driver`, `bdb-cli`) and then I want to sign the paylod if I feel everything is ok. Possibly on another machine where I store my private keys

File: docs/handcraft.rst [hunk 8]
```diff
@@ -0,0 +1,2049 @@
...
+
+    In [0]: alice.verifying_key
+
+We are now going to craft this payload by hand.
+
+Extract asset id and metadata id:
+
+.. ipython::
+
+    In [0]: asset_id = prepared_creation_tx['transaction']['asset']['id']
```
Comment (r-marques) [inline docs/handcraft.rst]: instead of doing this , maybe `str(uuid.uuid4())`.

Unless you want to end up with the exact same payload which is fine, but then just add a note about it. Or else it feels like cheating first creating a transaction with the driver and then using some of its values

File: docs/handcraft.rst [hunk 9]
```diff
@@ -0,0 +1,2049 @@
...
+       ...:             'serial_number': 'abcd1234',
+       ...:         },
+       ...:     },
+       ...:     'divisible': False,
+       ...:     'refillable': False,
+       ...:     'updatable': False,
+       ...:     'id': asset_id,
+       ...: }
+
+.. note:: In a signature-less system, the values for the keys:
```
Comment (r-marques) [inline docs/handcraft.rst]: I am not sure that this is worth mentioning since this is not going to happen. And its also bad practice because then you would have to trust the server not to change the payload

File: docs/handcraft.rst [hunk 10]
```diff
@@ -0,0 +1,2049 @@
...
+.. ipython::
+
+    In [0]: condition = {
+       ...:     'amount': 1,
+       ...:     'cid': 0,
+       ...:     'condition': {
+       ...:         'details': ed25519.to_dict(),
+       ...:         'uri': ed25519.condition_uri,
+       ...:     },
+       ...:     'owners_after': (alice.verifying_key,),
```
Comment (r-marques) [inline docs/handcraft.rst]: Does the serialize json replaces tuples with lists?

File: docs/handcraft.rst [hunk 11]
```diff
@@ -0,0 +1,2049 @@
...
+
+    In [0]: conditions = (condition,)
+
+The key part is the condition uri:
+
+.. ipython::
+
+    In [0]: ed25519.condition_uri
+
+.. todo:: Explain the basics of the above. cc, 4, 20, the hash of the key, 96
```
Comment (r-marques) [inline docs/handcraft.rst]: You can just point out that its part of the `cc` spec and point them to the ietf draft

File: docs/handcraft.rst (hunk 9 above)
Comment (sbellem) [inline docs/handcraft.rst]: Ah ok. The goal was to outline some key differences for developers who are familiar with HTTP server - client architectures in which the payloads can be very small since the server takes care of filling in default values.

But it can be removed.

File: docs/handcraft.rst (hunk 10 above)
Comment (sbellem) [inline docs/handcraft.rst]: Indeed! How cool is that?!

File: docs/handcraft.rst [hunk 12]
```diff
@@ -0,0 +1,2049 @@
...
+
+    In [0]: prepared_creation_tx_bk = deepcopy(prepared_creation_tx)
+
+    In [0]: # set fulfillment to None
+
+    In [0]: prepared_creation_tx['transaction']['fulfillments'][0]['fulfillment'] = None
+ 
+    In [0]: handcrafted_creation_tx == prepared_creation_tx
+
+Still not equal because we used tuples instead of strings
```
Comment (r-marques) [inline docs/handcraft.rst]: lists?

File: docs/handcraft.rst [hunk 13]
```diff
@@ -0,0 +1,2049 @@
...
+    import cryptoconditions
+    
+    from bigchaindb_driver.crypto import generate_keypair
+
+
+    alice = generate_keypair()
+
+    operation = 'CREATE'
+    
+    asset_id = str(uuid4())
```
Comment (r-marques) [inline docs/handcraft.rst]: Here it is. You may ignore my previous comment on the `uuid`

File: docs/handcraft.rst (hunk 8 above)
Comment (sbellem) [inline docs/handcraft.rst]: Yes, this is more like scaffolding that could be removed. What initially happened is that I pretty much reversed-engineered common, and wanted to double check that the hashes, and signatures were the same, hence re-using the asset id.

So what do you think, we could probably remove altogether the comparisons with the driver offchain tools (`prepare_transaction`, `fulfill_transaction`)?

File: docs/handcraft.rst [hunk 14]
```diff
@@ -0,0 +1,2049 @@
...
+
+    In [0]: prepared_transfer_tx_bk = deepcopy(prepared_transfer_tx)
+
+    In [0]: # set fulfillment to None
+
+    In [0]: prepared_transfer_tx['transaction']['fulfillments'][0]['fulfillment'] = None
+
+    In [0]: handcrafted_transfer_tx == prepared_transfer_tx
+
+Still not equal because we used tuples instead of strings
```
Comment (r-marques) [inline docs/handcraft.rst]: lists?

File: docs/handcraft.rst (hunk 7 above)
Comment (sbellem) [inline docs/handcraft.rst]: Ok, should I remove the entire comment? I do not mind. As far as the entire page is concerned, what really matter are the examples.

File: docs/handcraft.rst (hunk 13 above)
Comment (sbellem) [inline docs/handcraft.rst]: Well, yes, here because I did not care about comparing with the payloads generated by `prepare_transaction` and `fulfill_transaction`

File: docs/handcraft.rst (hunk 8 above)
Comment (r-marques) [inline docs/handcraft.rst]: No. After reading the entire document I like it this way. First you craft a transaction and compare it with what the driver outputs. And then you craft the same transaction without the support of the driver generating the uuids yourself.

I like it because if someone is going to develop a driver they can follow the same pattern and constantly compare it with a real transaction to make sure they are doing everything ok.

File: docs/handcraft.rst (hunk 9 above)
Comment (r-marques) [inline docs/handcraft.rst]: Oh. Then I think is better to just say that instead of using signature-less system

**Comment by r-marques** [comment]
I really like it. The language is simple and to the point and you almost make it sound simple
